### DISCUSSION TITLE
Performance between engine and provider

### DISCUSSION THREAD

**Discussion by Seulhui** [openssl/openssl#21910] [General Discussion]
Hello everyone,

I have my own crypto module and key storage, and to use it via OpenSSL, I used engine APIs.
Following the OpenSSL version upgrade to 3.0, I changed it to use provider API.

However, I can easily know the performance difference between the engine and the provider.
For your information, there is a wrapping API to call/use my own crypto and keystorage module and use it not only the engine but also the provider. So, only the OpenSSL engine and provider APIs are different.

It's more than 2-3 times slower even though I fetched the operations before. I also tested on OpenSSL 3.1.2, but it was a little faster than 3.0.7 but it was still much slower than the engine.

I wonder if this is the same for your side who implementing both engine and provider. If it is not the same, is it better to investigate more of the provider?

Regards,
Seulhui Lee.

**Comment by t8m** [comment]
From a performance perspective 3.1 should be significantly better than 3.0, and 3.2 should be a bit better than 3.1. There is further work planned for 3.3 (April 2024) which should hopefully make some additional improvements.

Performance issues are highly workload dependent so it depends exactly what you are doing to what extent you will encounter performance issues.

We are interested in building a set of "representative" performance workload tests to better target our performance improvement work. There are a few tests already here but we need more/better ones:

https://github.com/openssl/tools/tree/master/perf

If you have suggestions (or better yet PRs) to add to the tests there, that would be helpful.

**Comment by sharanakumarpatil** [comment]
Hello everyone, 
Is the performance of the provider better than that of the engine with the 3.5 version?
With respect to TLS handshake performance, is there any information available on this that would help?

**Reply by jericson** [reply]
Note that [the ENGINE API is deprecated in 3.0 and on](https://github.com/openssl/openssl/blob/master/README-ENGINES.md).

**Reply by sharanakumarpatil** [reply]
Thank you, we started using the provider, but we still observe engine performance is slightly better than the provider.
So, trying to figure out the current 3.5 version of the provider outperforms the engine, even compared to the engine's pre-deprecation performance.
